File: C/Recursion/TowerOfHanoi.c
#include <stdio.h>

// Function to move a disk from the source peg to the destination peg
void towerOfHanoi(int n, char source, char destination, char auxiliary)
{
    // Base case: If only one disk is present, move it directly from source to destination
    if (n == 1)
    {
        printf("Move disk 1 from peg %c to peg %c\n", source, destination);
        return;
    }

    // Move (n-1) disks from the source peg to the auxiliary peg
    towerOfHanoi(n - 1, source, auxiliary, destination);

    // Move the remaining disk from the source peg to the destination peg
    printf("Move disk %d from peg %c to peg %c\n", n, source, destination);

    // Move the (n-1) disks from the auxiliary peg to the destination peg
    towerOfHanoi(n - 1, auxiliary, destination, source);
}

int main()
{
    int numDisks;
    printf("Enter the number of disks: ");
    scanf("%d", &numDisks);

    printf("Tower of Hanoi solution:\n");
    towerOfHanoi(numDisks, 'A', 'C', 'B');

    return 0;
}

/*Explanation:

The towerOfHanoi function is a recursive function that takes four parameters:

n: the number of disks to be moved
source: the source peg from which the disks are initially placed
destination: the destination peg to which the disks are to be moved
auxiliary: the auxiliary peg used for temporarily holding the disks
The base case of the recursive function is when there is only one disk (n == 1). In this case, the function simply prints the move from the source peg to the destination peg.

For the case of more than one disk, the function follows the steps of the Tower of Hanoi algorithm:

Step 1: Move (n-1) disks from the source peg to the auxiliary peg using recursion. The destination peg is used as the auxiliary peg in this step.
Step 2: Move the remaining disk from the source peg to the destination peg directly.
Step 3: Move the (n-1) disks from the auxiliary peg to the destination peg using recursion. The source peg is used as the auxiliary peg in this step.
In the main function, the user is prompted to enter the number of disks they want to solve the Tower of Hanoi puzzle for.

The towerOfHanoi function is then called with the user-inputted number of disks, the source peg 'A', the destination peg 'C', and the auxiliary peg 'B'.

The function executes and prints each move required to solve the puzzle.

Finally, the main function returns 0, indicating successful execution of the program.
*/

File: Python/Recursion/TowerOfHanoi.py
"""
The Tower of Hanoi is a problem that involves moving n disks from one tower to another, 
following specific rules. 

It consists of three towers, usually labeled A, B, and C.

The puzzle starts with a number of disks of different sizes arranged in increasing order from top to bottom
(i.e. smallest on top and then so on...) on one tower, typically tower A. 

The objective is to move the entire stack of disks to another tower, following these rules:

Only one disk can be moved at a time.
Each move involves taking the top disk from one tower and placing it onto another tower.
The larger disk should never be placed on top of a smaller disk. The disks must always be arranged 
in decreasing order from bottom to top on any tower.

Algorithm: We will be using Recursion to solve this problem

S1: First we move top (n-1) disks from source(i.e. the starting tower) to helper(i.e. the middle tower)
S2: Now we are left with one disk at source which we will transfer directly to the destination(i.e. the last tower)
S3: Now we will move the (n-1) disk at the helper to the destination tower

"""


def display():
    num_disks = int(input("Enter the number of disks: "))
    process(num_disks,"A","B","C")
    #A, B, C are the names of the towers, starting from A and delivering all the disks to C



def process(n, src, helper, dest):
    if(n == 1): # base case
        print("Transfer disk" , n , "from" , src , "to" , dest)
        return 
    process(n-1,src,dest,helper)
    #moving the top (n-1) disks from A to B
    print("Transfer disk" , n , "from" , src , "to" , dest)
    #moving the remaining one disk from A to C
    process(n-1,helper,src,dest)
    #now moving the (n-1) disks on B to C

display()